Fails to build on macOS (10.14; AppleClang 10.0.0.10001145, boost 1.68)

macOS Mojave 10.14.3
The C compiler identification is AppleClang 10.0.0.10001145
The CXX compiler identification is AppleClang 10.0.0.10001145
Boost version: 1.68.0

boost and everything from homebrew, all up to date at the time of writing.


```
In file included from /Users/tko/Projects/ledger/system.hh:96:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/unordered_map:369:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__hash_table:19:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cmath:313:9: error:
      no member named 'signbit' in the global namespace
using ::signbit;
      ~~^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cmath:314:9: error:
      no member named 'fpclassify' in the global namespace
using ::fpclassify;
      ~~^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cmath:315:9: error:
      no member named 'isfinite' in the global namespace; did you mean 'finite'?
using ::isfinite;
      ~~^

[and more]

19 errors generated.
make[2]: *** [system.hh.pch] Error 1
make[1]: *** [src/CMakeFiles/libledger.dir/all] Error 2
make: *** [all] Error 2
acprep: ERROR: Execution failed: make
```

<details><p>

```console
% ./acprep make
acprep: INFO: Invoking primary phase: make
acprep: INFO: Executing phase: make
acprep: INFO: Executing phase: config
acprep: INFO: Executing phase: configure
acprep: INFO: System type is => Darwin
acprep: INFO: Setting up build flavor => debug
CMake Warning:
  No source or binary directory provided.  Both will be assumed to be the
  same as the current working directory, but note that this warning will
  become a fatal error in future CMake releases.


CMake Deprecation Warning at CMakeLists.txt:4 (cmake_policy):
  The OLD behavior for policy CMP0042 will be removed from a future version
  of CMake.

  The cmake-policies(7) manual explains that the OLD behaviors of all
  policies are deprecated and that a policy should be set to OLD only under
  specific short-term circumstances.  Projects should be ported to the NEW
  behavior and not rely on setting a policy to OLD.


-- The C compiler identification is AppleClang 10.0.0.10001145
-- The CXX compiler identification is AppleClang 10.0.0.10001145
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/usr/bin/g++
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/usr/bin/g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PythonInterp: /usr/local/bin/python2.7 (found version "2.7.16")
-- Boost version: 1.68.0
-- Found the following Boost libraries:
--   date_time
--   filesystem
--   system
--   iostreams
--   regex
--   unit_test_framework
-- Looking for getpwuid
-- Looking for getpwuid - found
-- Looking for getpwnam
-- Looking for getpwnam - found
-- Looking for ioctl
-- Looking for ioctl - found
-- Looking for isatty
-- Looking for isatty - found
-- Performing Test UNIX_PIPES_COMPILES
-- Performing Test UNIX_PIPES_COMPILES - Success
-- Performing Test BOOST_REGEX_UNICODE_RUNS
-- Performing Test BOOST_REGEX_UNICODE_RUNS - Failed
-- Looking for readline in edit
-- Looking for readline in edit - found
-- Performing Test HAVE_WORKING_UTFCPP
-- Performing Test HAVE_WORKING_UTFCPP - Success
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/tko/Projects/ledger
[  1%] Generating ../system.hh.pch
In file included from /Users/tko/Projects/ledger/system.hh:96:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/unordered_map:369:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__hash_table:19:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cmath:313:9: error:
      no member named 'signbit' in the global namespace
using ::signbit;
      ~~^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cmath:314:9: error:
      no member named 'fpclassify' in the global namespace
using ::fpclassify;
      ~~^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cmath:315:9: error:
      no member named 'isfinite' in the global namespace; did you mean 'finite'?
using ::isfinite;
      ~~^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/usr/include/math.h:749:12: note:
      'finite' declared here
extern int finite(double)
           ^
In file included from /Users/tko/Projects/ledger/system.hh:96:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/unordered_map:369:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__hash_table:19:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cmath:316:9: error:
      no member named 'isinf' in the global namespace
using ::isinf;
      ~~^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cmath:317:9: error:
      no member named 'isnan' in the global namespace
using ::isnan;
      ~~^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cmath:318:9: error:
      no member named 'isnormal' in the global namespace
using ::isnormal;
      ~~^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cmath:319:7: error:
      no member named 'isgreater' in the global namespace; did you mean
      '::std::greater'?
using ::isgreater;
      ^~
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/functional:720:29: note:
      '::std::greater' declared here
struct _LIBCPP_TEMPLATE_VIS greater : binary_function<_Tp, _Tp, bool>
                            ^
In file included from /Users/tko/Projects/ledger/system.hh:96:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/unordered_map:369:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__hash_table:19:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cmath:320:7: error:
      no member named 'isgreaterequal' in the global namespace; did you mean
      '::std::greater_equal'?
using ::isgreaterequal;
      ^~
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/functional:749:29: note:
      '::std::greater_equal' declared here
struct _LIBCPP_TEMPLATE_VIS greater_equal : binary_function<_Tp, _Tp, bool>
                            ^
In file included from /Users/tko/Projects/ledger/system.hh:96:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/unordered_map:369:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__hash_table:19:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cmath:321:9: error:
      no member named 'isless' in the global namespace
using ::isless;
      ~~^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cmath:322:9: error:
      no member named 'islessequal' in the global namespace
using ::islessequal;
      ~~^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cmath:323:9: error:
      no member named 'islessgreater' in the global namespace
using ::islessgreater;
      ~~^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cmath:324:9: error:
      no member named 'isunordered' in the global namespace
using ::isunordered;
      ~~^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cmath:325:9: error:
      no member named 'isunordered' in the global namespace
using ::isunordered;
      ~~^
In file included from /Users/tko/Projects/ledger/system.hh:141:
In file included from /usr/local/include/boost/date_time/posix_time/posix_time.hpp:24:
In file included from /usr/local/include/boost/date_time/posix_time/time_formatters.hpp:12:
In file included from /usr/local/include/boost/date_time/gregorian/gregorian.hpp:31:
In file included from /usr/local/include/boost/date_time/gregorian/gregorian_io.hpp:16:
In file included from /usr/local/include/boost/date_time/date_facet.hpp:23:
In file included from /usr/local/include/boost/date_time/period_parser.hpp:14:
In file included from /usr/local/include/boost/date_time/string_parse_tree.hpp:13:
In file included from /usr/local/include/boost/lexical_cast.hpp:32:
In file included from /usr/local/include/boost/lexical_cast/try_lexical_convert.hpp:42:
In file included from /usr/local/include/boost/lexical_cast/detail/converter_lexical.hpp:54:
In file included from /usr/local/include/boost/lexical_cast/detail/converter_lexical_streams.hpp:63:
In file included from /usr/local/include/boost/lexical_cast/detail/inf_nan.hpp:34:
/usr/local/include/boost/math/special_functions/sign.hpp:30:17: error: no member
      named 'signbit' in namespace 'std'; did you mean simply 'signbit'?
        return (std::signbit)(x) ? 1 : 0;
                ^~~~~
/usr/local/include/boost/math/special_functions/math_fwd.hpp:891:8: note:
      'signbit' declared here
   int signbit BOOST_NO_MACRO_EXPAND(T x);
       ^
In file included from /Users/tko/Projects/ledger/system.hh:141:
In file included from /usr/local/include/boost/date_time/posix_time/posix_time.hpp:24:
In file included from /usr/local/include/boost/date_time/posix_time/time_formatters.hpp:12:
In file included from /usr/local/include/boost/date_time/gregorian/gregorian.hpp:31:
In file included from /usr/local/include/boost/date_time/gregorian/gregorian_io.hpp:16:
In file included from /usr/local/include/boost/date_time/date_facet.hpp:23:
In file included from /usr/local/include/boost/date_time/period_parser.hpp:14:
In file included from /usr/local/include/boost/date_time/string_parse_tree.hpp:13:
In file included from /usr/local/include/boost/lexical_cast.hpp:32:
In file included from /usr/local/include/boost/lexical_cast/try_lexical_convert.hpp:42:
In file included from /usr/local/include/boost/lexical_cast/detail/converter_lexical.hpp:54:
In file included from /usr/local/include/boost/lexical_cast/detail/converter_lexical_streams.hpp:63:
In file included from /usr/local/include/boost/lexical_cast/detail/inf_nan.hpp:35:
/usr/local/include/boost/math/special_functions/fpclassify.hpp:155:12: error:
      no member named 'fpclassify' in namespace 'std'; did you mean simply
      'fpclassify'?
   return (std::fpclassify)(t);
           ^~~~~
/usr/local/include/boost/math/special_functions/math_fwd.hpp:876:8: note:
      'fpclassify' declared here
   int fpclassify BOOST_NO_MACRO_EXPAND(T t);
       ^
In file included from /Users/tko/Projects/ledger/system.hh:141:
In file included from /usr/local/include/boost/date_time/posix_time/posix_time.hpp:24:
In file included from /usr/local/include/boost/date_time/posix_time/time_formatters.hpp:12:
In file included from /usr/local/include/boost/date_time/gregorian/gregorian.hpp:31:
In file included from /usr/local/include/boost/date_time/gregorian/gregorian_io.hpp:16:
In file included from /usr/local/include/boost/date_time/date_facet.hpp:23:
In file included from /usr/local/include/boost/date_time/period_parser.hpp:14:
In file included from /usr/local/include/boost/date_time/string_parse_tree.hpp:13:
In file included from /usr/local/include/boost/lexical_cast.hpp:32:
In file included from /usr/local/include/boost/lexical_cast/try_lexical_convert.hpp:42:
In file included from /usr/local/include/boost/lexical_cast/detail/converter_lexical.hpp:54:
In file included from /usr/local/include/boost/lexical_cast/detail/converter_lexical_streams.hpp:63:
In file included from /usr/local/include/boost/lexical_cast/detail/inf_nan.hpp:35:
/usr/local/include/boost/math/special_functions/fpclassify.hpp:314:17: error:
      no member named 'isfinite' in namespace 'std'; did you mean simply
      'isfinite'?
        return (std::isfinite)(x);
                ^~~~~
/usr/local/include/boost/math/special_functions/math_fwd.hpp:879:9: note:
      'isfinite' declared here
   bool isfinite BOOST_NO_MACRO_EXPAND(T z);
        ^
In file included from /Users/tko/Projects/ledger/system.hh:141:
In file included from /usr/local/include/boost/date_time/posix_time/posix_time.hpp:24:
In file included from /usr/local/include/boost/date_time/posix_time/time_formatters.hpp:12:
In file included from /usr/local/include/boost/date_time/gregorian/gregorian.hpp:31:
In file included from /usr/local/include/boost/date_time/gregorian/gregorian_io.hpp:16:
In file included from /usr/local/include/boost/date_time/date_facet.hpp:23:
In file included from /usr/local/include/boost/date_time/period_parser.hpp:14:
In file included from /usr/local/include/boost/date_time/string_parse_tree.hpp:13:
In file included from /usr/local/include/boost/lexical_cast.hpp:32:
In file included from /usr/local/include/boost/lexical_cast/try_lexical_convert.hpp:42:
In file included from /usr/local/include/boost/lexical_cast/detail/converter_lexical.hpp:54:
In file included from /usr/local/include/boost/lexical_cast/detail/converter_lexical_streams.hpp:63:
In file included from /usr/local/include/boost/lexical_cast/detail/inf_nan.hpp:35:
/usr/local/include/boost/math/special_functions/fpclassify.hpp:385:17: error:
      no member named 'isnormal' in namespace 'std'; did you mean simply
      'isnormal'?
        return (std::isnormal)(x);
                ^~~~~
/usr/local/include/boost/math/special_functions/math_fwd.hpp:888:9: note:
      'isnormal' declared here
   bool isnormal BOOST_NO_MACRO_EXPAND(T t);
        ^
In file included from /Users/tko/Projects/ledger/system.hh:141:
In file included from /usr/local/include/boost/date_time/posix_time/posix_time.hpp:24:
In file included from /usr/local/include/boost/date_time/posix_time/time_formatters.hpp:12:
In file included from /usr/local/include/boost/date_time/gregorian/gregorian.hpp:31:
In file included from /usr/local/include/boost/date_time/gregorian/gregorian_io.hpp:16:
In file included from /usr/local/include/boost/date_time/date_facet.hpp:23:
In file included from /usr/local/include/boost/date_time/period_parser.hpp:14:
In file included from /usr/local/include/boost/date_time/string_parse_tree.hpp:13:
In file included from /usr/local/include/boost/lexical_cast.hpp:32:
In file included from /usr/local/include/boost/lexical_cast/try_lexical_convert.hpp:42:
In file included from /usr/local/include/boost/lexical_cast/detail/converter_lexical.hpp:54:
In file included from /usr/local/include/boost/lexical_cast/detail/converter_lexical_streams.hpp:63:
In file included from /usr/local/include/boost/lexical_cast/detail/inf_nan.hpp:35:
/usr/local/include/boost/math/special_functions/fpclassify.hpp:456:17: error:
      no member named 'isinf' in namespace 'std'; did you mean simply 'isinf'?
        return (std::isinf)(x);
                ^~~~~
/usr/local/include/boost/math/special_functions/math_fwd.hpp:882:9: note:
      'isinf' declared here
   bool isinf BOOST_NO_MACRO_EXPAND(T t);
        ^
In file included from /Users/tko/Projects/ledger/system.hh:141:
In file included from /usr/local/include/boost/date_time/posix_time/posix_time.hpp:24:
In file included from /usr/local/include/boost/date_time/posix_time/time_formatters.hpp:12:
In file included from /usr/local/include/boost/date_time/gregorian/gregorian.hpp:31:
In file included from /usr/local/include/boost/date_time/gregorian/gregorian_io.hpp:16:
In file included from /usr/local/include/boost/date_time/date_facet.hpp:23:
In file included from /usr/local/include/boost/date_time/period_parser.hpp:14:
In file included from /usr/local/include/boost/date_time/string_parse_tree.hpp:13:
In file included from /usr/local/include/boost/lexical_cast.hpp:32:
In file included from /usr/local/include/boost/lexical_cast/try_lexical_convert.hpp:42:
In file included from /usr/local/include/boost/lexical_cast/detail/converter_lexical.hpp:54:
In file included from /usr/local/include/boost/lexical_cast/detail/converter_lexical_streams.hpp:63:
In file included from /usr/local/include/boost/lexical_cast/detail/inf_nan.hpp:35:
/usr/local/include/boost/math/special_functions/fpclassify.hpp:552:17: error:
      no member named 'isnan' in namespace 'std'; did you mean simply 'isnan'?
        return (std::isnan)(x);
                ^~~~~
/usr/local/include/boost/math/special_functions/math_fwd.hpp:885:9: note:
      'isnan' declared here
   bool isnan BOOST_NO_MACRO_EXPAND(T t);
        ^
19 errors generated.
make[2]: *** [system.hh.pch] Error 1
make[1]: *** [src/CMakeFiles/libledger.dir/all] Error 2
make: *** [all] Error 2
acprep: ERROR: Execution failed: make
```
</p></details>

--- a/.travis.yml
+++ b/.travis.yml
@@ -57,8 +57,8 @@ addons:
       - mpfr
 
 before_script:
-  # On macOS, precompiling system.hh does not work, and boost-python packaging is broken
-  - if [ "$TRAVIS_OS_NAME" = osx ]; then EXTRA_CMAKE_ARGS="-DPRECOMPILE_SYSTEM_HH=OFF -DBoost_NO_BOOST_CMAKE=ON"; fi
+  # On macOS boost-python packaging is broken
+  - if [ "$TRAVIS_OS_NAME" = osx ]; then EXTRA_CMAKE_ARGS="-DBoost_NO_BOOST_CMAKE=ON"; fi
   - cmake . -DUSE_PYTHON=ON -DPython_FIND_VERSION_MAJOR=${PY_MAJOR} -DBUILD_DEBUG=ON $EXTRA_CMAKE_ARGS
   - make VERBOSE=1
 

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -265,7 +265,8 @@ else()
   endmacro(ADD_PCH_RULE _header_filename _src_list _other_srcs)
 endif()
 
-if(PRECOMPILE_SYSTEM_HH)
+if(PRECOMPILE_SYSTEM_HH AND NOT (COMMAND target_precompile_headers))
+  # enable fallback for CMake versions older than 3.16 without target_precompile_headers
   add_pch_rule(${PROJECT_BINARY_DIR}/system.hh LEDGER_SOURCES LEDGER_CLI_SOURCES)
 endif()
 
@@ -292,6 +293,15 @@ if (BUILD_LIBRARY)
 else()
   add_executable(ledger ${LEDGER_SOURCES} main.cc global.cc)
   add_ledger_library_dependencies(ledger)
+endif()
+
+if (PRECOMPILE_SYSTEM_HH AND (COMMAND target_precompile_headers) AND (CMAKE_BUILD_TYPE STREQUAL "Debug"))
+  if (BUILD_LIBRARY)
+    target_precompile_headers(libledger PRIVATE ${PROJECT_BINARY_DIR}/system.hh)
+    target_precompile_headers(ledger REUSE_FROM libledger)
+  else()
+    target_precompile_headers(ledger PRIVATE ${PROJECT_BINARY_DIR}/system.hh)
+  endif()
 endif()
 
 if (USE_PYTHON)

